File: app/src/main/java/com/soundscore/app/ui/screens/LogScreen.kt
package com.soundscore.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.soundscore.app.data.model.Album
import com.soundscore.app.ui.components.AlbumArtwork
import com.soundscore.app.ui.components.GlassCard
import com.soundscore.app.ui.components.ScreenHeader
import com.soundscore.app.ui.components.SectionHeader
import com.soundscore.app.ui.components.StarRating
import com.soundscore.app.ui.components.StatPill
import com.soundscore.app.ui.components.SyncBanner
import com.soundscore.app.ui.components.TimelineEntry
import com.soundscore.app.ui.theme.AccentAmber
import com.soundscore.app.ui.theme.AccentGreen
import com.soundscore.app.ui.theme.AccentGreenDim
import com.soundscore.app.ui.theme.ChromeLight
import com.soundscore.app.ui.theme.DarkBase
import com.soundscore.app.ui.theme.DarkElevated
import com.soundscore.app.ui.theme.FeedItemBorder
import com.soundscore.app.ui.theme.GlassBg
import com.soundscore.app.ui.theme.GlassBorder
import com.soundscore.app.ui.theme.TextSecondary
import com.soundscore.app.ui.theme.TextTertiary
import com.soundscore.app.ui.viewmodel.LogUiState
import com.soundscore.app.ui.viewmodel.LogViewModel
import com.soundscore.app.ui.viewmodel.RecentLogEntry

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LogScreen(
    modifier: Modifier = Modifier,
    logViewModel: LogViewModel = viewModel(),
) {
    val uiState by logViewModel.uiState.collectAsStateWithLifecycle()
    var showSearchSheet by remember { mutableStateOf(false) }

    if (showSearchSheet) {
        ModalBottomSheet(
            onDismissRequest = { showSearchSheet = false },
            sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true),
            containerColor = DarkElevated,
            shape = RoundedCornerShape(topStart = 24.dp, topEnd = 24.dp),
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 24.dp, vertical = 16.dp),
            ) {
                Text(
                    text = "Log an album",
                    style = MaterialTheme.typography.headlineSmall,
                    color = ChromeLight,
                )
                Spacer(Modifier.height(16.dp))
                Text(
                    text = "Album search coming soon. Use the Quick Rate cards below to log albums.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = TextSecondary,
                )
                Spacer(Modifier.height(24.dp))
            }
        }
    }

    Box(modifier = modifier.fillMaxSize()) {
        LogScreenContent(
            uiState = uiState,
            onRate = logViewModel::updateRating,
        )

        FloatingActionButton(
            onClick = { showSearchSheet = true },
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(end = 20.dp, bottom = 24.dp),
            shape = CircleShape,
            containerColor = AccentGreen,
            contentColor = DarkBase,
        ) {
            Icon(Icons.Filled.Add, contentDescription = "Log Album")
        }
    }
}

@Composable
fun LogScreenContent(
    uiState: LogUiState,
    onRate: (String, Float) -> Unit,
    modifier: Modifier = Modifier,
) {
    LazyColumn(
        modifier = modifier.fillMaxSize(),
        contentPadding = PaddingValues(start = 20.dp, top = 16.dp, end = 20.dp, bottom = 120.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        item {
            SyncBanner(message = uiState.syncMessage)
        }

        item {
            ScreenHeader(
                title = "Diary",
                subtitle = "Your listening journal. Rate, log, repeat.",
            )
        }

        item {
            GlassCard(
                cornerRadius = 22.dp,
                borderColor = FeedItemBorder,
                frosted = true,
            ) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceEvenly,
                ) {
                    uiState.summaryStats.forEach { stat ->
                        Column(horizontalAlignment = Alignment.CenterHorizontally) {
                            Text(
                                text = stat.value,
                                style = MaterialTheme.typography.headlineMedium,
                                color = if (stat.label == "This week") AccentGreen else ChromeLight,
                                fontWeight = FontWeight.Black,
                            )
                            Spacer(Modifier.height(2.dp))
                            Text(
                                text = stat.label.uppercase(),
                                style = MaterialTheme.typography.labelSmall,
                                color = TextTertiary,
                            )
                        }
                    }
                }
            }
        }

        item {
            SectionHeader(
                eyebrow = "Quick rate",
                title = "Tap to rate",
            )
        }

        item {
            LazyRow(
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                contentPadding = PaddingValues(end = 8.dp),
            ) {
                items(uiState.quickLogAlbums, key = { it.id }) { album ->
                    QuickRateCard(
                        album = album,
                        rating = uiState.ratings[album.id] ?: 0f,
                        onRate = { onRate(album.id, it) },
                    )
                }
            }
        }

        if (uiState.recentLogs.isNotEmpty()) {
            item {
                SectionHeader(
                    eyebrow = "Recent",
                    title = "Your diary entries",
                )
            }

            items(uiState.recentLogs, key = { "${it.album.id}-${it.timeLabel}" }) { entry ->
                TimelineEntry(
                    dateLabel = entry.dateLabel,
                    timeLabel = entry.timeLabel,
                ) {
                    DiaryEntryCard(entry = entry)
                }
            }
        }

        item {
            GlassCard(
                cornerRadius = 20.dp,
                borderColor = FeedItemBorder,
            ) {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 8.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                ) {
                    Text(
                        text = "Write Later",
                        style = MaterialTheme.typography.titleMedium,
                        color = ChromeLight,
                    )
                    Spacer(Modifier.height(4.dp))
                    Text(
                        text = "Queue albums for later review — coming soon",
                        style = MaterialTheme.typography.bodySmall,
                        color = TextTertiary,
                    )
                }
            }
        }
    }
}

@Composable
private fun QuickRateCard(
    album: Album,
    rating: Float,
    onRate: (Float) -> Unit,
) {
    GlassCard(
        modifier = Modifier.width(140.dp),
        cornerRadius = 20.dp,
        fillMaxWidth = false,
        borderColor = FeedItemBorder,
        contentPadding = PaddingValues(8.dp),
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Box {
                AlbumArtwork(
                    artworkUrl = album.artworkUrl,
                    colors = album.artColors,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(130.dp),
                    cornerRadius = 14.dp,
                )
                if (rating > 0f) {
                    Box(
                        modifier = Modifier
                            .align(Alignment.TopEnd)
                            .padding(6.dp)
                            .clip(RoundedCornerShape(10.dp))
                            .background(DarkBase.copy(alpha = 0.7f))
                            .padding(horizontal = 6.dp, vertical = 3.dp),
                    ) {
                        Text(
                            text = String.format("%.1f", rating),
                            style = MaterialTheme.typography.labelSmall,
                            color = AccentAmber,
                            fontWeight = FontWeight.Bold,
                        )
                    }
                }
            }
            Text(
                text = album.title,
                style = MaterialTheme.typography.titleSmall,
                maxLines = 1,
                fontWeight = FontWeight.SemiBold,
            )
            Text(
                text = album.artist,
                style = MaterialTheme.typography.bodySmall,
                color = TextSecondary,
                maxLines = 1,
            )
            StarRating(
                rating = rating,
                onRate = onRate,
                starSize = 14.dp,
            )
        }
    }
}

@Composable
private fun DiaryEntryCard(entry: RecentLogEntry) {
    GlassCard(
        cornerRadius = 18.dp,
        borderColor = FeedItemBorder,
        contentPadding = PaddingValues(10.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            AlbumArtwork(
                artworkUrl = entry.album.artworkUrl,
                colors = entry.album.artColors,
                modifier = Modifier.size(56.dp),
                cornerRadius = 14.dp,
            )
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                Text(
                    text = entry.album.title,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                )
                Text(
                    text = entry.album.artist,
                    style = MaterialTheme.typography.bodySmall,
                    color = TextSecondary,
                )
                if (entry.caption.isNotBlank()) {
                    Text(
                        text = entry.caption,
                        style = MaterialTheme.typography.bodySmall,
                        color = TextTertiary,
                        maxLines = 1,
                    )
                }
            }
            StarRating(rating = entry.rating, starSize = 12.dp)
        }
    }
}

File: ios/SoundScore/SoundScore/Screens/LogScreen.swift
import SwiftUI

struct LogScreen: View {
    @StateObject private var viewModel = LogViewModel()
    var onSelectAlbum: (Album) -> Void = { _ in }
    @State private var showSearchSheet = false
    @State private var diaryMode: Int = 0

    var body: some View {
        ZStack(alignment: .bottomTrailing) {
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 16) {
                    SyncBanner(message: viewModel.syncMessage)

                    if let error = viewModel.errorMessage {
                        ErrorBanner(message: error, onRetry: {
                            Task { await SoundScoreRepository.shared.refresh() }
                        })
                    }

                    ScreenHeader(title: "Diary", subtitle: "Your listening journal. Rate, log, repeat.")

                    // Albums/Songs toggle
                    GlassSegmentedControl(items: ["Albums", "Songs"], selection: $diaryMode)

                    if !viewModel.summaryStats.isEmpty {
                        HStack(spacing: 10) {
                            ForEach(Array(viewModel.summaryStats.enumerated()), id: \.offset) { index, stat in
                                let gradients: [[Color]] = [AlbumColors.forest, AlbumColors.amber, AlbumColors.orchid]
                                VStack(spacing: 4) {
                                    Text(stat.value)
                                        .font(.system(size: 22, weight: .black, design: .rounded))
                                        .foregroundStyle(
                                            LinearGradient(
                                                colors: gradients[index % gradients.count],
                                                startPoint: .topLeading, endPoint: .bottomTrailing
                                            )
                                        )
                                    Text(stat.label.uppercased())
                                        .font(SSTypography.labelSmall)
                                        .foregroundColor(SSColors.textTertiary)
                                        .fontWeight(.semibold)
                                }
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 14)
                                .background(SSColors.glassBg)
                                .clipShape(RoundedRectangle(cornerRadius: 16))
                                .overlay(
                                    RoundedRectangle(cornerRadius: 16)
                                        .stroke(gradients[index % gradients.count].last!.opacity(0.25), lineWidth: 1)
                                )
                            }
                        }
                    }

                    if diaryMode == 0 {
                        // Albums mode
                        SectionHeader(eyebrow: "Quick rate", title: "Tap to rate")

                        ScrollView(.horizontal, showsIndicators: false) {
                            LazyHStack(spacing: 12) {
                                ForEach(viewModel.quickLogAlbums) { album in
                                    QuickRateCard(
                                        album: album,
                                        rating: viewModel.ratings[album.id] ?? 0,
                                        onRate: { viewModel.updateRating(albumId: album.id, rating: $0) },
                                        onSelectAlbum: onSelectAlbum
                                    )
                                }
                            }
                            .padding(.trailing, 8)
                        }

                        if !viewModel.recentLogs.isEmpty {
                            SectionHeader(eyebrow: "Recent", title: "Your diary entries")

                            ForEach(viewModel.recentLogs) { entry in
                                TimelineEntry(dateLabel: entry.dateLabel, timeLabel: entry.timeLabel) {
                                    DiaryEntryCard(entry: entry, onSelectAlbum: onSelectAlbum)
                                }
                            }
                        }
                    } else {
                        // Songs mode
                        SectionHeader(eyebrow: "Track log", title: "Recently rated songs")

                        ForEach(viewModel.recentSongLogs) { entry in
                            SongLogEntryCard(entry: entry, onSelectAlbum: onSelectAlbum)
                        }

                        if viewModel.recentSongLogs.isEmpty {
                            EmptyState(
                                title: "No song ratings yet",
                                subtitle: "Open an album and rate individual tracks to see them here.",
                                icon: "music.note"
                            )
                        }
                    }

                    // Write Later — hidden until feature ships
                }
                .padding(.horizontal, 20)
                .padding(.top, 16)
                .padding(.bottom, 120)
            }
            .refreshable { await SoundScoreRepository.shared.refresh() }

            Button {
                UIImpactFeedbackGenerator(style: .medium).impactOccurred()
                showSearchSheet = true
            } label: {
                Image(systemName: "plus")
                    .font(.system(size: 22, weight: .bold))
                    .foregroundColor(SSColors.darkBase)
                    .frame(width: 56, height: 56)
                    .background(ThemeManager.shared.primary)
                    .clipShape(Circle())
                    .shadow(color: ThemeManager.shared.primary.opacity(0.3), radius: 10, y: 4)
            }
            .padding(.trailing, 20)
            .padding(.bottom, 100)
        }
        .sheet(isPresented: $showSearchSheet) {
            QuickLogSearchSheet(onSelectAlbum: { album in
                showSearchSheet = false
                onSelectAlbum(album)
            })
            .presentationDetents([.large])
            .presentationDragIndicator(.visible)
            .presentationBackground(SSColors.darkElevated)
        }
    }
}

// MARK: - Quick Log Search Sheet

private struct QuickLogSearchSheet: View {
    @StateObject private var viewModel = SearchViewModel()
    var onSelectAlbum: (Album) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Log an album")
                .font(SSTypography.headlineSmall)
                .foregroundColor(SSColors.chromeLight)

            PillSearchBar(query: $viewModel.query, placeholder: "Search albums to log...")

            if viewModel.results.isEmpty && !viewModel.query.isEmpty {
                EmptyState(
                    title: "No matches",
                    subtitle: "Try a different search term.",
                    icon: "magnifyingglass"
                )
            } else {
                ScrollView {
                    LazyVStack(spacing: 8) {
                        ForEach(viewModel.results) { album in
                            Button {
                                UIImpactFeedbackGenerator(style: .light).impactOccurred()
                                onSelectAlbum(album)
                            } label: {
                                HStack(spacing: 12) {
                                    AlbumArtwork(artworkUrl: album.artworkUrl, colors: album.artColors, cornerRadius: 12)
                                        .frame(width: 48, height: 48)
                                    VStack(alignment: .leading, spacing: 2) {
                                        Text(album.title)
                                            .font(SSTypography.titleMedium)
                                            .foregroundColor(SSColors.chromeLight)
                                            .fontWeight(.semibold)
                                        Text(album.artist)
                                            .font(SSTypography.bodySmall)
                                            .foregroundColor(SSColors.textSecondary)
                                    }
                                    Spacer()
                                    Image(systemName: "chevron.right")
                                        .font(.system(size: 12))
                                        .foregroundColor(SSColors.chromeFaint)
                                }
                                .padding(.vertical, 4)
                            }
                            .buttonStyle(.plain)
                        }
                    }
                }
            }

            Spacer()
        }
        .padding(.horizontal, 24)
        .padding(.top, 24)
    }
}

private struct QuickRateCard: View {
    let album: Album
    let rating: Float
    let onRate: (Float) -> Void
    var onSelectAlbum: (Album) -> Void = { _ in }

    private var dominantColor: Color {
        album.artColors.first ?? ThemeManager.shared.primary
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            ZStack(alignment: .topTrailing) {
                AlbumArtwork(artworkUrl: album.artworkUrl, colors: album.artColors, cornerRadius: 18)
                    .frame(width: 148, height: 148)
                    .onTapGesture {
                        UIImpactFeedbackGenerator(style: .light).impactOccurred()
                        onSelectAlbum(album)
                    }

                if rating > 0 {
                    Text(String(format: "%.1f", rating))
                        .font(.system(size: 12, weight: .heavy, design: .rounded))
                        .foregroundColor(.white)
                        .padding(.horizontal, 8)
                        .padding(.vertical, 4)
                        .background(dominantColor.opacity(0.85))
                        .clipShape(Capsule())
                        .padding(8)
                        .transition(.scale.combined(with: .opacity))
                }
            }

            VStack(alignment: .leading, spacing: 4) {
                Text(album.title)
                    .font(SSTypography.titleMedium)
                    .fontWeight(.bold)
                    .foregroundColor(SSColors.chromeLight)
                    .lineLimit(1)
                Text(album.artist)
                    .font(SSTypography.bodySmall)
                    .foregroundColor(SSColors.textSecondary)
                    .lineLimit(1)
                StarRating(rating: rating, onRate: onRate, starSize: 14)
            }
            .padding(.horizontal, 4)
            .padding(.top, 8)
            .padding(.bottom, 4)
        }
        .frame(width: 148)
    }
}

private struct SongLogEntryCard: View {
    let entry: RecentSongLogEntry
    var onSelectAlbum: (Album) -> Void = { _ in }

    var body: some View {
        GlassCard(cornerRadius: 18, borderColor: SSColors.feedItemBorder,
                  contentPadding: EdgeInsets(top: 10, leading: 10, bottom: 10, trailing: 10)) {
            HStack(spacing: 10) {
                AlbumArtwork(artworkUrl: entry.album.artworkUrl, colors: entry.album.artColors, cornerRadius: 12)
                    .frame(width: 48, height: 48)
                    .onTapGesture {
                        UIImpactFeedbackGenerator(style: .light).impactOccurred()
                        onSelectAlbum(entry.album)
                    }

                VStack(alignment: .leading, spacing: 2) {
                    Text(entry.track.title)
                        .font(SSTypography.titleMedium)
                        .fontWeight(.semibold)
                        .foregroundColor(SSColors.chromeLight)
                        .lineLimit(1)
                    Text("\(entry.album.artist) · \(entry.album.title)")
                        .font(SSTypography.bodySmall)
                        .foregroundColor(SSColors.textSecondary)
                        .lineLimit(1)
                }

                Spacer()

                VStack(alignment: .trailing, spacing: 3) {
                    StarRating(rating: entry.rating, starSize: 11)
                    Text(entry.dateLabel)
                        .font(SSTypography.labelSmall)
                        .foregroundColor(SSColors.textTertiary)
                }
            }
        }
    }
}

private struct DiaryEntryCard: View {
    let entry: RecentLogEntry
    var onSelectAlbum: (Album) -> Void = { _ in }

    var body: some View {
        GlassCard(cornerRadius: 18, borderColor: SSColors.feedItemBorder, contentPadding: EdgeInsets(top: 10, leading: 10, bottom: 10, trailing: 10)) {
            HStack(spacing: 10) {
                AlbumArtwork(artworkUrl: entry.album.artworkUrl, colors: entry.album.artColors, cornerRadius: 14)
                    .frame(width: 56, height: 56)
                    .onTapGesture {
                        UIImpactFeedbackGenerator(style: .light).impactOccurred()
                        onSelectAlbum(entry.album)
                    }
                VStack(alignment: .leading, spacing: 2) {
                    Text(entry.album.title)
                        .font(SSTypography.titleMedium)
                        .fontWeight(.semibold)
                        .foregroundColor(SSColors.chromeLight)
                    Text(entry.album.artist)
                        .font(SSTypography.bodySmall)
                        .foregroundColor(SSColors.textSecondary)
                    if !entry.caption.isEmpty {
                        Text(entry.caption)
                            .font(SSTypography.bodySmall)
                            .foregroundColor(SSColors.textTertiary)
                            .lineLimit(1)
                    }
                }
                Spacer()
                StarRating(rating: entry.rating, starSize: 12)
            }
        }
    }
}
